### PR TITLE
Support Aggregating Multiple Kubernetes Clusters into a Single xDS Cluster

### PR DESCRIPTION
**Pull request by minwoox**
Motivation:
The current `XdsKubernetesService` does not allow combining multiple Kubernetes clusters into a single xDS cluster.

Modifications:
- Introduced `KubernetesEndpointAggregator` to support aggregation of multiple Kubernetes clusters into one xDS cluster.
  - The fields in `KubernetesEndpointAggregator` are similar to `ClusterLoadAssignment`, with the addition of `ServiceEndpointWatcher` to retrieve Kubernetes information from multiple clusters.

Result:
- xDS now supports the aggregation of multiple Kubernetes clusters.

### CODE REVIEW

File: xds/src/main/proto/centraldogma/xds/k8s/v1/xds_kubernetes.proto [hunk 1]
```diff
@@ -19,70 +19,89 @@ option java_multiple_files = true;
...
 
 service XdsKubernetesService {
 
-  rpc CreateServiceEndpointWatcher(CreateServiceEndpointWatcherRequest) returns (ServiceEndpointWatcher) {
+  rpc CreateKubernetesEndpointAggregator(CreateKubernetesEndpointAggregatorRequest)
+      returns (KubernetesEndpointAggregator) {
     option (google.api.http) = {
-      post: "/api/v1/xds/{parent=groups/*}/k8s/watchers"
-      body: "watcher"
+      post: "/api/v1/xds/{parent=groups/*}/k8s/endpointAggregators"
```
Comment (minwoox) [inline xds/src/main/proto/centraldogma/xds/k8s/v1/xds_kubernetes.proto:36]: Used camel case for the collection identifiers:
https://github.com/aip-dev/google.aip.dev/blob/master/aip/general/0122.md#collection-identifiers

File: xds/src/main/java/com/linecorp/centraldogma/xds/k8s/v1/XdsKubernetesService.java [hunk 2]
```diff
@@ -78,81 +82,103 @@ public XdsKubernetesService(XdsResourceManager xdsResourceManager) {
...
-                        Status.INTERNAL.withDescription(
-                                "Failed to retrieve k8s endpoints within 5 seconds. watcherName: " +
-                                watcher.getName()).asRuntimeException());
-            }, 5, TimeUnit.SECONDS);
-
-            whenReady.handle((endpoints, cause1) -> {
-                if (!completed.compareAndSet(false, true)) {
+        final ArrayList<CompletableFuture<Void>> futures = new ArrayList<>();
+        for (KubernetesLocalityLbEndpoints kubernetesLocalityLbEndpoints
+                : aggregator.getLocalityLbEndpointsList()) {
```
Comment (ikhoon) [inline xds/src/main/java/com/linecorp/centraldogma/xds/k8s/v1/XdsKubernetesService.java:122]: What do you think of returning`INVALID_ARGUMENT` if `aggregator.getLocalityLbEndpointsList().size() == 0`?

**Review by ikhoon — COMMENTED** [review]
Very nice. Left minor comments.

File: xds/src/main/proto/centraldogma/xds/k8s/v1/xds_kubernetes.proto [hunk 3]
```diff
@@ -19,70 +19,89 @@ option java_multiple_files = true;
...
-message DeleteServiceEndpointWatcherRequest {
-  string name = 1 [(google.api.field_behavior) = IDENTIFIER];
+message ServiceEndpointWatcher {
+  string service_name = 1 [(google.api.field_behavior) = REQUIRED];
+  string port_name = 2 [(google.api.field_behavior) = OPTIONAL];
+  Kubeconfig kubeconfig = 3 [(google.api.field_behavior) = REQUIRED];
 }
 
-message KubernetesConfig {
+message Kubeconfig {
```
[the remark this answers is not in the document]
Comment (minwoox) [reply xds/src/main/proto/centraldogma/xds/k8s/v1/xds_kubernetes.proto:88]: Isn't `Kubeconfig` one word?
https://kubernetes.io/docs/concepts/configuration/organize-cluster-access-kubeconfig/

File: xds/src/main/java/com/linecorp/centraldogma/xds/k8s/v1/XdsKubernetesEndpointFetchingService.java [hunk 4]
```diff
@@ -109,80 +112,81 @@ protected ScheduledExecutorService executor() {
...
-            }, true);
+                new KubernetesEndpointsUpdater(commandExecutor, futures, executorService,
+                                               groupName, aggregator);
+        updaters.put(aggregatorName, updater);
+        CompletableFuture.allOf(futures.toArray(EMPTY_FUTURES)).exceptionally(cause -> {
+            logger.warn("Unexpected exception while creating a KubernetesEndpointGroup in fetching service",
+                        cause);
+            // Do not remove the updater from updaters because it can remove the updater that is created
+            // by the next commit. The updater will be removed only when the file or group is removed.
+            updater.close();
```
[the remark this answers is not in the document]
Comment (minwoox) [reply xds/src/main/java/com/linecorp/centraldogma/xds/k8s/v1/XdsKubernetesEndpointFetchingService.java:159]: If we remove the updater, there's a chance that the updater, which is created by the next commit, is removed. So I think it's safe to remove the updater only when the corresponding file is removed.

File: xds/src/main/java/com/linecorp/centraldogma/xds/k8s/v1/XdsKubernetesEndpointFetchingService.java [hunk 5]
```diff
@@ -213,88 +217,79 @@ protected boolean isStopped() {
...
+            }));
         }
 
-        void maybeSchedule(KubernetesEndpointGroup kubernetesEndpointGroup) {
-            if (scheduledFuture != null) {
+        void maybeSchedule() {
+            if (closing || scheduledFuture != null) {
                 return;
             }
             // Commit after 1 second so that it pushes with all the fetched endpoints in the duration
```
[the remark this answers is not in the document]
Comment (minwoox) [reply xds/src/main/java/com/linecorp/centraldogma/xds/k8s/v1/XdsKubernetesEndpointFetchingService.java]: Forgot to update it. 😓

File: xds/src/main/java/com/linecorp/centraldogma/xds/k8s/v1/XdsKubernetesService.java (hunk 2 above)
Comment (minwoox) [reply xds/src/main/java/com/linecorp/centraldogma/xds/k8s/v1/XdsKubernetesService.java]: That's a good suggestion. 👍 👍

File: xds/src/main/proto/centraldogma/xds/k8s/v1/xds_kubernetes.proto (hunk 3 above)
[the remark this answers is not in the document]
Comment (ikhoon) [reply xds/src/main/proto/centraldogma/xds/k8s/v1/xds_kubernetes.proto:88]: It is. I only checked how the fabric kubernetes client named a class for `kubeconfig`.
Let's keep it as is.

**Review by ikhoon — APPROVED** [review]
Thanks! 🙏🙏

**Pull request merged**